File: app/src/main/java/moe/yuuta/mipushtester/log/LogUtils.java
package moe.yuuta.mipushtester.log;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;

import java.io.File;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.FileProvider;
import moe.yuuta.mipushtester.BuildConfig;

import static android.content.Intent.EXTRA_STREAM;

public class LogUtils {
    public static String getLogFolder (@NonNull Context context) {
        return context.getCacheDir().getPath() + "/logs";
    }

    @Nullable
    public static Intent getShareIntent(Context context) {
        File zipFile = new File(context.getExternalCacheDir().getAbsolutePath() + "/logs/logs-" +
                new SimpleDateFormat("yyyy-mm-dd-H-m-s", Locale.US).format(new Date()) + ".zip");
        try {
            com.elvishew.xlog.LogUtils.compress(getLogFolder(context),
                    zipFile.getAbsolutePath());
            Uri fileUri = FileProvider.getUriForFile(
                    context,
                    BuildConfig.APPLICATION_ID + ".fileprovider",
                    zipFile);
            if (fileUri == null || !zipFile.exists()) {
                throw new NullPointerException();
            }
            Intent intent = new Intent();
            intent.setAction(Intent.ACTION_SEND);
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
            String type = context.getContentResolver().getType(fileUri);
            if (type == null || type.trim().equals("")) {
                type = "application/zip";
            }
            intent.setType(type);
            intent.putExtra(EXTRA_STREAM, fileUri);
            return intent;
        } catch (IOException | NullPointerException e) {
            // TODO: Maybe some error are occurred?
            return null;
        }
    }
}

File: app/src/main/java/moe/yuuta/mipushtester/log/LogUtils.kt
package moe.yuuta.mipushtester.log

import android.content.Context
import android.content.Intent
import android.content.Intent.EXTRA_STREAM
import android.util.Log
import androidx.annotation.NonNull
import androidx.annotation.Nullable
import androidx.core.content.FileProvider
import com.elvishew.xlog.XLog
import moe.yuuta.mipushtester.BuildConfig
import java.io.File
import java.text.SimpleDateFormat
import java.util.*

object LogUtils {
    fun getLogFolder(@NonNull context: Context): String =
            context.cacheDir.path + "/logs"

    @Nullable
    fun getShareIntent(context: Context): Intent? {
        val zipFile = File("${context.externalCacheDir.absolutePath}/logs/logs-" +
                "${SimpleDateFormat("yyyy-mm-dd-H-m-s", Locale.US).format(Date())}.zip")
        try {
            com.elvishew.xlog.LogUtils.compress(getLogFolder(context),
                    zipFile.absolutePath)
            val fileUri = FileProvider.getUriForFile(
                    context,
                    BuildConfig.APPLICATION_ID + ".fileprovider",
                    zipFile)
            if (fileUri == null || !zipFile.exists()) {
                throw NullPointerException()
            }
            val intent = Intent()
            intent.action = Intent.ACTION_SEND
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            var type = context.contentResolver.getType(fileUri)
            if (type == null || type.trim().equals("")) {
                type = "application/zip"
            }
            intent.type = type
            intent.putExtra(EXTRA_STREAM, fileUri)
            return intent
        } catch (e: Exception) {
            try {
                XLog.tag(LogUtils::class.simpleName).build()
                        .e("Share logs", e)
            } catch (ignored: Exception) {}
            System.err.println("Unable to share logs, ${Log.getStackTraceString(e)}")
            return null
        }
    }
}
